parser/docker_cmd.go

```diff
@@ -114,7 +120,151 @@ func (b *BuildCmd) Run() (string, error) {
 			"error":    nil,
 		},
 	})
-	return image.ID, nil
+
+	var retID = image.ID
+
+	if opts.Squash {
+		libsquash.Verbose = true
+
+		var imageReader1, pipeWriter1 = io.Pipe()
+		var imageReader2, pipeWriter2 = io.Pipe()
+		var squashedImageReader, squashedImageWriter = io.Pipe()
+
+		defer func() {
+			for _, pipeReader := range []*io.PipeReader{imageReader1, imageReader2, squashedImageReader} {
+				pipeReader.Close()
+			}
+
+			for _, pipeWriter := range []*io.PipeWriter{pipeWriter1, pipeWriter2, squashedImageWriter} {
+				pipeWriter.Close()
+			}
+		}()
+
+		// exporting in a goroutine since the io.Pipe is synchronous
+		go func() {
```

Comment:
Definitely. This is all experimental right now - my main focus is reducing
the memory usage of the squash functionality. I'll clean all this up
afterwards.

On Sunday, January 4, 2015, Cole Gleason <notifications@github.com> wrote:

> In parser/docker_cmd.go
> <https://github.com/winchman/builder-core/pull/22#discussion-diff-22442342>
> :
>
> > +		var imageReader1, pipeWriter1 = io.Pipe()
> > +		var imageReader2, pipeWriter2 = io.Pipe()
> > +		var squashedImageReader, squashedImageWriter = io.Pipe()
> > +
> > +		defer func() {
> > +			for _, pipeReader := range []*io.PipeReader{imageReader1, imageReader2, squashedImageReader} {
> > +				pipeReader.Close()
> > +			}
> > +
> > +			for _, pipeWriter := range []*io.PipeWriter{pipeWriter1, pipeWriter2, squashedImageWriter} {
> > +				pipeWriter.Close()
> > +			}
> > +		}()
> > +
> > +		// exporting in a goroutine since the io.Pipe is synchronous
> > +		go func() {
>
> Is it possible to break these goroutine functions out of the Run() command?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/winchman/builder-core/pull/22/files#r22442342>.
>